flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NettyShuffleEnvironment.java

```diff
@@ -228,91 +211,111 @@ public void releasePartitions(Collection<ResultPartitionID> partitionIds) {
-			String taskName,
-			ExecutionAttemptID executionId,
-			Collection<ResultPartitionDeploymentDescriptor> resultPartitionDeploymentDescriptors,
-			MetricGroup outputGroup,
-			MetricGroup buffersGroup) {
+	@Override
+	public ShuffleIOOwnerContext createShuffleIOOwnerContext(
+			String ownerName,
+			ExecutionAttemptID executionAttemptID,
+			MetricGroup parentGroup) {
+		MetricGroup nettyGroup = createShuffleIOOwnerMetricGroup(checkNotNull(parentGroup));
+		return new ShuffleIOOwnerContext(
+			checkNotNull(ownerName),
+			checkNotNull(executionAttemptID),
+			parentGroup,
+			nettyGroup.addGroup(METRIC_GROUP_INPUT),
+			nettyGroup.addGroup(METRIC_GROUP_OUTPUT));
+	}
+
+	@Override
+	public Collection<ResultPartition> createResultPartitionWriters(
+			ShuffleIOOwnerContext ownerContext,
+			Collection<ResultPartitionDeploymentDescriptor> resultPartitionDeploymentDescriptors) {
 		synchronized (lock) {
-			Preconditions.checkState(!isShutdown, "The NetworkEnvironment has already been shut down.");
+			Preconditions.checkState(!isClosed, "The NettyShuffleEnvironment has already been shut down.");
 
 			ResultPartition[] resultPartitions = new ResultPartition[resultPartitionDeploymentDescriptors.size()];
 			int counter = 0;
 			for (ResultPartitionDeploymentDescriptor rpdd : resultPartitionDeploymentDescriptors) {
-				resultPartitions[counter++] = resultPartitionFactory.create(taskName, executionId, rpdd);
+				resultPartitions[counter++] = resultPartitionFactory.create(
+					ownerContext.getOwnerName(),
+					ownerContext.getExecutionAttemptID(),
+					rpdd);
 			}
 
-			registerOutputMetrics(outputGroup, buffersGroup, resultPartitions);
-			return resultPartitions;
+			MetricGroup outputMetricGroup = ownerContext.getOutputGroup();
+			registerOutputMetrics(
+				config.isNetworkDetailedMetrics(),
+				outputMetricGroup,
+				outputMetricGroup.addGroup(METRIC_GROUP_BUFFERS),
+				resultPartitions);
+			return  Arrays.asList(resultPartitions);
 		}
 	}
 
-	public SingleInputGate[] createInputGates(
-			String taskName,
-			ExecutionAttemptID executionId,
+	@Override
+	public Collection<SingleInputGate> createInputGates(
+			ShuffleIOOwnerContext ownerContext,
 			PartitionProducerStateProvider partitionProducerStateProvider,
-			Collection<InputGateDeploymentDescriptor> inputGateDeploymentDescriptors,
-			MetricGroup parentGroup,
-			MetricGroup inputGroup,
-			MetricGroup buffersGroup) {
+			Collection<InputGateDeploymentDescriptor> inputGateDeploymentDescriptors) {
 		synchronized (lock) {
-			Preconditions.checkState(!isShutdown, "The NetworkEnvironment has already been shut down.");
+			Preconditions.checkState(!isClosed, "The NettyShuffleEnvironment has already been shut down.");
+
+			MetricGroup networkInputGroup = ownerContext.getInputGroup();
+			@SuppressWarnings("deprecation")
+			InputChannelMetrics inputChannelMetrics = new InputChannelMetrics(networkInputGroup, ownerContext.getParentGroup());
```

Comment:
Maybe we could delay the creation of `InputChannelMetrics` in the `SingleInputGateFactory#createInputChannels`, then the following would be 
`SingleInputGate inputGate = singleInputGateFactory.create(ownerContext, icdd, partitionProducerStateProvider)` instead. 
To do so we could make the current `createInputGates` simple and make all the things done when required.

Reply:
The gate creation loop will create `InputChannelMetrics` multiple times, I would avoid this because `InputChannelMetrics` are shared by gates at the moment.